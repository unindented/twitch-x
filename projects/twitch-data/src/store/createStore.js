import {
  applyMiddleware,
  combineReducers,
  createStore
} from 'redux'
import thunkMiddleware from 'redux-thunk'
import {composeWithDevTools} from 'redux-devtools-extension'
import * as reducers from '../reducers'

export default function createEnhancedStore (options = {}) {
  const otherReducers = options.reducers || {}
  const otherMiddleware = options.middleware || []
  const otherEnhancers = options.enhancers || []

  const dataReducer = combineReducers(reducers)
  const allReducers = combineReducers({...otherReducers, data: dataReducer})

  const allMiddleware = otherMiddleware
    .concat(thunkMiddleware)

  const allEnhancers = otherEnhancers
    .concat(applyMiddleware(...allMiddleware))

  return createStore(allReducers, composeWithDevTools(...allEnhancers))
}
