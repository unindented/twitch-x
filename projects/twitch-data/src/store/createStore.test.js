import createStore from './createStore'

describe('store/createStore', () => {
  describe('with no options', () => {
    let store

    beforeEach(() => {
      store = createStore()
    })

    it('creates a store with no additional reducers', () => {
      const state = store.getState()
      expect(state).toMatchSnapshot()
    })

    it('types the state properly', () => {
      const state = store.getState()
      expect(state.data.count.value).toBe(0)
    })
  })

  describe('with reducers', () => {
    let store

    beforeEach(() => {
      const reducers = {
        foo: (state = {bar: 'baz'}, action) => {
          switch (action.type) {
            case 'QUX':
              return {...state, qux: 'quux'}
          }
          return state
        }
      }
      store = createStore({reducers})
    })

    it('creates a store with the additional reducers', () => {
      expect(store.getState()).toMatchSnapshot()
    })

    it('processes actions', () => {
      const action = {type: 'QUX'}
      store.dispatch(action)

      const state = store.getState()
      expect(state).toMatchSnapshot()
    })

    it('processes thunk actions', () => {
      const action = {type: 'QUX'}
      const thunkAction = () => (dispatch) => { dispatch(action) }
      store.dispatch(thunkAction())

      const state = store.getState()
      expect(state).toMatchSnapshot()
    })

    it('types the state properly', () => {
      const state = store.getState()
      expect(state.data.count.value).toBe(0)
      expect(state.foo.bar).toBe('baz')
    })
  })
})
