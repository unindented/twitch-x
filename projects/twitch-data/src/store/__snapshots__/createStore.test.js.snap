// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store/createStore with no options creates a store with no additional reducers 1`] = `
Object {
  "data": Object {
    "channels": Object {
      "byId": Object {},
      "search": Array [],
    },
    "communities": Object {
      "byId": Object {},
      "top": Array [],
    },
    "games": Object {
      "byId": Object {},
      "search": Array [],
      "top": Array [],
    },
    "session": Object {
      "token": null,
    },
    "streams": Object {
      "byId": Object {},
      "search": Array [],
      "top": Array [],
    },
  },
}
`;

exports[`store/createStore with reducers creates a store with the additional reducers 1`] = `
Object {
  "data": Object {
    "channels": Object {
      "byId": Object {},
      "search": Array [],
    },
    "communities": Object {
      "byId": Object {},
      "top": Array [],
    },
    "games": Object {
      "byId": Object {},
      "search": Array [],
      "top": Array [],
    },
    "session": Object {
      "token": null,
    },
    "streams": Object {
      "byId": Object {},
      "search": Array [],
      "top": Array [],
    },
  },
  "foo": Object {
    "bar": "baz",
  },
}
`;

exports[`store/createStore with reducers processes actions 1`] = `
Object {
  "data": Object {
    "channels": Object {
      "byId": Object {},
      "search": Array [],
    },
    "communities": Object {
      "byId": Object {},
      "top": Array [],
    },
    "games": Object {
      "byId": Object {},
      "search": Array [],
      "top": Array [],
    },
    "session": Object {
      "token": null,
    },
    "streams": Object {
      "byId": Object {},
      "search": Array [],
      "top": Array [],
    },
  },
  "foo": Object {
    "bar": "baz",
    "qux": "quux",
  },
}
`;

exports[`store/createStore with reducers processes thunk actions 1`] = `
Object {
  "data": Object {
    "channels": Object {
      "byId": Object {},
      "search": Array [],
    },
    "communities": Object {
      "byId": Object {},
      "top": Array [],
    },
    "games": Object {
      "byId": Object {},
      "search": Array [],
      "top": Array [],
    },
    "session": Object {
      "token": null,
    },
    "streams": Object {
      "byId": Object {},
      "search": Array [],
      "top": Array [],
    },
  },
  "foo": Object {
    "bar": "baz",
    "qux": "quux",
  },
}
`;
